### PR TITLE
Update Jest definitions with more inline documentation

### PR DESCRIPTION
**Pull request by orta**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- x ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: http://facebook.github.io/jest/docs/en/mock-function-api.html#mockfnmockclear
- [x] Increase the version number in the header if appropriate.

No major changes, I did add `mockRestore` though - http://facebook.github.io/jest/docs/en/mock-function-api.html#mockfnmockrestore

### CODE REVIEW

**Comment by orta** [comment]
This is green now 👍

File: types/jest/index.d.ts [hunk 1]
```diff
@@ -119,13 +120,24 @@ declare namespace jest {
...
+        /** Notes that this async test can run concurrently. */
         concurrent: It;
     }
 
     interface Describe {
+        /** 
+         * Creates a scope for holding tests
+         * 
+         * @param {string} name The name of your scope
+         * @param {fn?} ProvidesCallback The function which provides the scope to work in
```
Comment (jwbay) [inline types/jest/index.d.ts]: ProvidesCallback is a little out of place here. Should this be `@param {Function} fn`?

File: types/jest/index.d.ts [hunk 2]
```diff
@@ -315,17 +328,79 @@ declare namespace jest {
...
+         */
+        mockRestore(): void;
+        /**
+         * Accepts a function that should be used as the implementation of the mock. 
+         * 
+         * The mock itself will still record all calls that go into and instances that come from itself – the only difference is that the implementation will also be executed when the mock is called.
+         * 
+         * Note: jest.fn(implementation) is a shorthand for mockImplementation.
+         * 
+         * @param {Function} fn The initola implmentation
```
Comment (jwbay) [inline types/jest/index.d.ts]: typo, initial?

**Review by jwbay — CHANGES_REQUESTED** [review]

File: types/jest/index.d.ts [hunk 3]
```diff
@@ -119,13 +120,24 @@ declare namespace jest {
...
         skip: It;
+        /** Notes that this async test can run concurrently. */
         concurrent: It;
     }
 
     interface Describe {
+        /** 
+         * Creates a scope for holding tests
+         * 
+         * @param {string} name The name of your scope
```
Comment (rbuckton) [inline types/jest/index.d.ts]: JSDoc types like `{string}` aren't necessary in a TypeScript declaration file.

File: types/jest/index.d.ts (hunk 1 above)
Comment (rbuckton) [inline types/jest/index.d.ts]: Or just `@param fn` as the JSDoc type is not necessary.

File: types/jest/index.d.ts (hunk 2 above)
Comment (rbuckton) [inline types/jest/index.d.ts]: Also, as above the `{Function}` annotation is not very useful here.

File: types/jest/index.d.ts [hunk 4]
```diff
@@ -315,17 +328,79 @@ declare namespace jest {
...
+         * @example
+         * var myMockFn = jest.fn(() => 'default')
+         *   .mockImplementationOnce(() => 'first call')
+         *   .mockImplementationOnce(() => 'second call');
+
+         * console.log(myMockFn(), myMockFn(), myMockFn(), myMockFn());
+         * > 'first call', 'second call', 'default', 'default'
+         * 
+         * @param {Function} fn A factory function to run the next time a mock is called
+         * @returns {Mock<T>} The same Jest mock
```
Comment (rbuckton) [inline types/jest/index.d.ts]: These JSDoc types are not necessary in a .d.ts file.

File: types/jest/index.d.ts [hunk 5]
```diff
@@ -315,17 +328,79 @@ declare namespace jest {
...
+         */
         mockImplementationOnce(fn: Function): Mock<T>;
+        /**
+         * Just a simple sugar function for: `jest.fn( () => this)`.
+         * 
+         * @returns {Mock<T>} The same Jest mock
```
Comment (rbuckton) [inline types/jest/index.d.ts]: JSDoc type `{Mock<T>}` not necessary in a .d.ts file.

**Review by rbuckton — CHANGES_REQUESTED** [review]

**Comment by vkrol** [comment]
I think that we need to get rid of this deprecation warning too https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/jest/index.d.ts#L37
Also we need to fix the description of the resetAllMocks method https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/jest/index.d.ts#L40

File: types/jest/index.d.ts [hunk 6]
```diff
@@ -315,17 +328,79 @@ declare namespace jest {
...
+         */
         mockReturnThis(): Mock<T>;
+        /**
+         * @deprecated Use ` jest.fn(() => value` instead.
```
Comment (jwbay) [inline types/jest/index.d.ts]: I was able to get some clarification (via https://github.com/facebook/jest/pull/3993 being merged) that this is not actually deprecated. The documentation for `mockReturnValueOnce` is updated as well. https://github.com/facebook/jest/blob/master/docs/en/MockFunctionAPI.md#mockfnmockreturnvaluevalue

**Comment by RyanCavanaugh** [comment]
@orta Please address comments from the code reviewers.

**Comment by orta** [comment]
Oops, yeah, I missed all these - sorry! I think it's better that #18034 gets merged, I'll re-create this later

**Pull request closed without merging**